File: Engr-T.stark/uniquePaths.cpp
class Solution {

public:

    vector<vector<int>> dp;

int uniquePathsUtil(int m, int n) {

    if(m == 1 || n == 1) return 1;

    if(dp[m][n] != -1)

        return dp[m][n];

    return dp[m][n] = uniquePathsUtil(m-1, n) + uniquePathsUtil(m, n-1);

}

int uniquePaths(int m, int n) {

    dp.clear();

    dp.assign(m+1, vector<int>(n+1, -1));

    return uniquePathsUtil(m, n);

}

};

File: Richie/Unique paths/unique-paths.py
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        return self.dfs(0, 0, m, n, {})
        
    def dfs(self, row, col, m, n, memo):
        id = col + row * n
        
        if row == m-1 and col == n-1:
            memo[id] = 1
        elif row >= m or col >= n:
            memo[id] = 0
        
        if id not in memo:    
            memo[id] = self.dfs(row+1, col, m, n, memo) + self.dfs(row, col+1, m, n, memo)
        return memo[id]
